runtime/src/accounts_db.rs

```diff
@@ -2396,21 +2403,34 @@ impl AccountsDB {
         );
     }
 
-    fn purge_slot_cache_keys(&self, dead_slot: Slot, slot_cache: SlotCache) {
-        // Slot purged from cache should not exist in the backing store
-        assert!(self.storage.get_slot_stores(dead_slot).is_none());
+    fn purge_slot_cache(&self, purge_slot: Slot, slot_cache: SlotCache) -> Vec<(u64, AccountInfo)> {
         let mut purged_slot_pubkeys: HashSet<(Slot, Pubkey)> = HashSet::new();
         let pubkey_to_slot_set: Vec<(Pubkey, Slot)> = slot_cache
             .iter()
             .map(|account| {
-                purged_slot_pubkeys.insert((dead_slot, *account.key()));
-                (*account.key(), dead_slot)
+                purged_slot_pubkeys.insert((purge_slot, *account.key()));
+                (*account.key(), purge_slot)
             })
             .collect();
+        self.purge_slot_cache_keys(purge_slot, purged_slot_pubkeys, pubkey_to_slot_set, true)
+    }
+
+    fn purge_slot_cache_keys(
+        &self,
+        dead_slot: Slot,
+        dead_slot_pubkeys: HashSet<(Slot, Pubkey)>,
+        pubkey_to_slot_set: Vec<(Pubkey, Slot)>,
+        is_dead: bool,
+    ) -> Vec<(u64, AccountInfo)> {
```

Comment:
Third charm. :)

nits: You don't need to return this, right?

Reply:
arg yeah, this keeps creeping in here, thanks!